File: frontend/src/api/governance.ts
import { governedFetch } from "./context";
import { API_BASE_URL } from "./health";

export interface Tenant { id: number; code: string; name: string; display_name: string; status: string; default_currency: string; timezone: string; created_at: string; }
export interface Member { id: number; tenant_id: number; user_id: number; user_email: string; user_display_name: string; status: string; roles: string[]; joined_at: string | null; }
export interface User { id: number; email: string; display_name: string; status: string; }
export interface Role { id: number; code: string; name: string; scope: string; }
export interface AuditEvent { id: number; occurred_at: string; actor_user_id: number | null; actor_type: string; action: string; entity_type: string; entity_id: string | null; description: string; pipeline_run_id: number | null; source_file_id: number | null; event_type: string; metadata_json: Record<string, unknown> | null; changes: Array<{field_name: string; old_value_json: unknown; new_value_json: unknown}>; }
interface Page<T> { items: T[]; total: number; }

async function parse<T>(response: Response): Promise<T> {
  const body = (await response.json()) as T & { detail?: string };
  if (!response.ok) throw new Error(body.detail ?? `Request failed (${response.status})`);
  return body;
}

export async function fetchTenants(): Promise<Tenant[]> { return parse<Tenant[]>(await governedFetch(`${API_BASE_URL}/tenants`)); }
export async function createTenant(values: Omit<Tenant, "id" | "created_at" | "status">): Promise<Tenant> {
  return parse<Tenant>(await governedFetch(`${API_BASE_URL}/tenants`, { method: "POST", headers: {"Content-Type": "application/json"}, body: JSON.stringify({...values, status: "active", fiscal_year_start_month: 1}) }));
}
export async function updateTenant(id: number, values: Partial<Tenant>): Promise<Tenant> { return parse<Tenant>(await governedFetch(`${API_BASE_URL}/tenants/${id}`, {method: "PATCH", headers: {"Content-Type": "application/json"}, body: JSON.stringify(values)})); }
export async function setTenantArchived(id: number, archive: boolean): Promise<Tenant> { return parse<Tenant>(await governedFetch(`${API_BASE_URL}/tenants/${id}/${archive ? "archive" : "restore"}`, {method: "POST"})); }
export async function fetchMembers(tenantId: number): Promise<Member[]> { return parse<Member[]>(await governedFetch(`${API_BASE_URL}/tenants/${tenantId}/members`)); }
export async function fetchUsers(): Promise<User[]> { return parse<User[]>(await governedFetch(`${API_BASE_URL}/users`)); }
export async function fetchRoles(): Promise<Role[]> { return parse<Role[]>(await governedFetch(`${API_BASE_URL}/roles`)); }
export async function addMember(tenantId: number, userId: number): Promise<Member> { return parse<Member>(await governedFetch(`${API_BASE_URL}/tenants/${tenantId}/members`, {method: "POST", headers: {"Content-Type": "application/json"}, body: JSON.stringify({user_id: userId, status: "active"})})); }
export async function assignMemberRole(membershipId: number, roleId: number): Promise<void> { const response = await governedFetch(`${API_BASE_URL}/tenant-memberships/${membershipId}/roles`, {method: "POST", headers: {"Content-Type": "application/json"}, body: JSON.stringify({role_id: roleId})}); if (!response.ok) throw new Error(`Request failed (${response.status})`); }
export async function removeMemberRole(membershipId: number, roleId: number): Promise<void> { const response = await governedFetch(`${API_BASE_URL}/tenant-memberships/${membershipId}/roles/${roleId}`, {method: "DELETE"}); if (!response.ok) throw new Error(`Request failed (${response.status})`); }
export async function fetchAuditEvents(action = "", entityType = "", occurredFrom = "", occurredTo = ""): Promise<AuditEvent[]> {
  const params = new URLSearchParams({page_size: "100"}); if (action) params.set("action", action); if (entityType) params.set("entity_type", entityType); if (occurredFrom) params.set("occurred_from", new Date(occurredFrom).toISOString()); if (occurredTo) params.set("occurred_to", new Date(`${occurredTo}T23:59:59`).toISOString());
  return (await parse<Page<AuditEvent>>(await governedFetch(`${API_BASE_URL}/audit-events?${params}`))).items;
}

File: backend/app/models/governance.py
from datetime import datetime
from typing import Any

from sqlalchemy import (
    JSON,
    Boolean,
    DateTime,
    ForeignKey,
    Integer,
    String,
    Text,
    UniqueConstraint,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class Tenant(Base):
    __tablename__ = "tenants"
    id: Mapped[int] = mapped_column(primary_key=True)
    code: Mapped[str] = mapped_column(String(100), unique=True, index=True)
    name: Mapped[str] = mapped_column(String(255))
    legal_name: Mapped[str | None] = mapped_column(String(255))
    display_name: Mapped[str] = mapped_column(String(255))
    status: Mapped[str] = mapped_column(String(30), index=True)
    default_currency: Mapped[str] = mapped_column(String(3))
    timezone: Mapped[str] = mapped_column(String(100))
    fiscal_year_start_month: Mapped[int] = mapped_column(Integer())
    data_retention_days: Mapped[int | None] = mapped_column(Integer())
    settings_json: Mapped[dict[str, Any] | None] = mapped_column(JSON())
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )
    archived_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))


class User(Base):
    __tablename__ = "users"
    id: Mapped[int] = mapped_column(primary_key=True)
    email: Mapped[str] = mapped_column(String(320), unique=True, index=True)
    display_name: Mapped[str] = mapped_column(String(255))
    first_name: Mapped[str | None] = mapped_column(String(100))
    last_name: Mapped[str | None] = mapped_column(String(100))
    status: Mapped[str] = mapped_column(String(30), index=True)
    password_hash: Mapped[str | None] = mapped_column(String(500))
    is_platform_admin: Mapped[bool] = mapped_column(Boolean(), default=False)
    last_login_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )
    archived_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))


class Role(Base):
    __tablename__ = "roles"
    id: Mapped[int] = mapped_column(primary_key=True)
    code: Mapped[str] = mapped_column(String(100), unique=True, index=True)
    name: Mapped[str] = mapped_column(String(255))
    description: Mapped[str | None] = mapped_column(Text())
    scope: Mapped[str] = mapped_column(String(30), index=True)
    is_system_role: Mapped[bool] = mapped_column(Boolean(), default=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )
    permissions: Mapped[list["RolePermission"]] = relationship(cascade="all, delete-orphan")


class Permission(Base):
    __tablename__ = "permissions"
    id: Mapped[int] = mapped_column(primary_key=True)
    code: Mapped[str] = mapped_column(String(120), unique=True, index=True)
    name: Mapped[str] = mapped_column(String(255))
    description: Mapped[str | None] = mapped_column(Text())
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )


class TenantUser(Base):
    __tablename__ = "tenant_users"
    __table_args__ = (UniqueConstraint("tenant_id", "user_id", name="uq_tenant_user"),)
    id: Mapped[int] = mapped_column(primary_key=True)
    tenant_id: Mapped[int] = mapped_column(
        ForeignKey("tenants.id", ondelete="RESTRICT"), index=True
    )
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="RESTRICT"), index=True)
    status: Mapped[str] = mapped_column(String(30), index=True)
    invited_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    joined_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    suspended_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )
    tenant: Mapped[Tenant] = relationship()
    user: Mapped[User] = relationship()
    role_assignments: Mapped[list["TenantUserRole"]] = relationship(cascade="all, delete-orphan")


class RolePermission(Base):
    __tablename__ = "role_permissions"
    __table_args__ = (UniqueConstraint("role_id", "permission_id", name="uq_role_permission"),)
    id: Mapped[int] = mapped_column(primary_key=True)
    role_id: Mapped[int] = mapped_column(ForeignKey("roles.id", ondelete="CASCADE"), index=True)
    permission_id: Mapped[int] = mapped_column(
        ForeignKey("permissions.id", ondelete="CASCADE"), index=True
    )
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    permission: Mapped[Permission] = relationship()


class TenantUserRole(Base):
    __tablename__ = "tenant_user_roles"
    __table_args__ = (UniqueConstraint("tenant_user_id", "role_id", name="uq_tenant_user_role"),)
    id: Mapped[int] = mapped_column(primary_key=True)
    tenant_user_id: Mapped[int] = mapped_column(
        ForeignKey("tenant_users.id", ondelete="CASCADE"), index=True
    )
    role_id: Mapped[int] = mapped_column(ForeignKey("roles.id", ondelete="RESTRICT"), index=True)
    assigned_by_user_id: Mapped[int | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL")
    )
    assigned_at: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    role: Mapped[Role] = relationship()


class AuditEvent(Base):
    __tablename__ = "audit_events"
    id: Mapped[int] = mapped_column(primary_key=True)
    tenant_id: Mapped[int | None] = mapped_column(
        ForeignKey("tenants.id", ondelete="RESTRICT"), index=True
    )
    actor_user_id: Mapped[int | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), index=True
    )
    actor_type: Mapped[str] = mapped_column(String(30))
    event_type: Mapped[str] = mapped_column(String(100), index=True)
    entity_type: Mapped[str] = mapped_column(String(100), index=True)
    entity_id: Mapped[str | None] = mapped_column(String(100), index=True)
    action: Mapped[str] = mapped_column(String(50), index=True)
    description: Mapped[str] = mapped_column(Text())
    request_id: Mapped[str | None] = mapped_column(String(100))
    correlation_id: Mapped[str | None] = mapped_column(String(100))
    pipeline_run_id: Mapped[int | None] = mapped_column(
        ForeignKey("pipeline_runs.id", ondelete="SET NULL"), index=True
    )
    source_file_id: Mapped[int | None] = mapped_column(
        ForeignKey("source_files.id", ondelete="SET NULL"), index=True
    )
    ip_address: Mapped[str | None] = mapped_column(String(64))
    user_agent: Mapped[str | None] = mapped_column(String(500))
    metadata_json: Mapped[dict[str, Any] | None] = mapped_column(JSON())
    occurred_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    changes: Mapped[list["AuditEventChange"]] = relationship(cascade="all, delete-orphan")


class AuditEventChange(Base):
    __tablename__ = "audit_event_changes"
    id: Mapped[int] = mapped_column(primary_key=True)
    audit_event_id: Mapped[int] = mapped_column(
        ForeignKey("audit_events.id", ondelete="CASCADE"), index=True
    )
    field_name: Mapped[str] = mapped_column(String(255))
    old_value_json: Mapped[Any | None] = mapped_column(JSON())
    new_value_json: Mapped[Any | None] = mapped_column(JSON())
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


class PipelineDefinition(Base):
    __tablename__ = "pipeline_definitions"
    __table_args__ = (UniqueConstraint("code", "version", name="uq_pipeline_definition_version"),)
    id: Mapped[int] = mapped_column(primary_key=True)
    code: Mapped[str] = mapped_column(String(100), index=True)
    name: Mapped[str] = mapped_column(String(255))
    description: Mapped[str | None] = mapped_column(Text())
    version: Mapped[str] = mapped_column(String(30))
    is_active: Mapped[bool] = mapped_column(Boolean(), default=True, index=True)
    configuration_schema_json: Mapped[dict[str, Any] | None] = mapped_column(JSON())
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )


class PipelineRunArtifact(Base):
    __tablename__ = "pipeline_run_artifacts"
    id: Mapped[int] = mapped_column(primary_key=True)
    tenant_id: Mapped[int] = mapped_column(
        ForeignKey("tenants.id", ondelete="RESTRICT"), index=True
    )
    pipeline_run_id: Mapped[int] = mapped_column(
        ForeignKey("pipeline_runs.id", ondelete="CASCADE"), index=True
    )
    artifact_type: Mapped[str] = mapped_column(String(50), index=True)
    name: Mapped[str] = mapped_column(String(255))
    relative_path: Mapped[str] = mapped_column(String(500))
    checksum: Mapped[str | None] = mapped_column(String(64))
    mime_type: Mapped[str | None] = mapped_column(String(255))
    file_size_bytes: Mapped[int | None] = mapped_column(Integer())
    metadata_json: Mapped[dict[str, Any] | None] = mapped_column(JSON())
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
